nupic/embodied/envs/rollout.py

```diff
@@ -0,0 +1,294 @@
+from collections import deque, defaultdict
+
+import numpy as np
+
+
+class Rollout(object):
+    """Collect rollouts of experiences in the environments and process them.
+
+    Parameters
+    ----------
+    ob_space : Space
+        Observation space properties (from env.observation_space).
+    ac_space : Space
+        Action space properties (from env.action_space).
+    nenvs : int
+        Number of environments used for collecting experiences.
+    nsteps_per_seg : int
+        Number of steps per rollout segment in each environment. (~like batch size?)
+    nsegs_per_env : int
+        Number of segments per environment in a rollout..
+    nlumps : type
+        ..
+    envs : [VecEnv]
+        List of VecEnvs to use for experience collection.
+    policy : object
+        CnnPolicy used for action selection.
+    int_rew_coeff : float
+        Coefficient for the internal reward (disagreement).
+    ext_rew_coeff : float
+        Coefficient for the external reward from the environment.
+    dynamics_list : [object]
+        List of dynamics networks.
+
+    Attributes:
+    ----------
+    nsteps : int
+        nsteps_per_seg * nsegs_per_env.
+    lump_stride : int
+        TODO.
+    reward_fun : lambda
+        reward function specifying how to combine internal and external rewards.
+    buf_vpreds : array
+        Buffer of value estimates.
+    buf_nlps : array
+        Buffer of negative log probabilities.
+    buf_rews : array
+        Buffer of rewards.
+    buf_ext_rews : array
+        Buffer of external rewards.
+    buf_acs : array
+        Buffer of actions.
+    buf_obs : array
+        Buffer of observations.
+    buf_obs_last : array
+        Buffer of last observations.
+    buf_news : array
+        Buffer of 'dones'.
+    buf_new_last : array
+        Buffer of last 'dones'.
+    buf_vpred_last : array
+        Buffer of last value estimates.
+    env_results : List
+        Outputs from the environment (observations, rewards, done, info).
+    int_rew : array
+        Internal rewards.
+    statlists : dict
+        Dictionary with run statistics.
+
+    """
+
+    def __init__(
+        self,
+        ob_space,
+        ac_space,
+        nenvs,
+        nsteps_per_seg,
+        nsegs_per_env,
+        nlumps,
+        envs,
+        policy,
+        int_rew_coeff,
+        ext_rew_coeff,
+        dynamics_list,
+    ):
+        self.nenvs = nenvs
+        self.nsteps_per_seg = nsteps_per_seg
+        self.nsegs_per_env = nsegs_per_env
+        self.nsteps = self.nsteps_per_seg * self.nsegs_per_env
+        self.ob_space = ob_space
+        self.ac_space = ac_space
+        self.nlumps = nlumps
+        self.lump_stride = nenvs // self.nlumps
+        self.envs = envs
+        self.policy = policy
+        self.dynamics_list = dynamics_list
+
+        # Define the reward function as a weighted combination of internal and (clipped)
+        # external rewards.
+        self.reward_fun = (
```

Comment:
I am not a big fan of using lambdas for currying. The reason is if you redefine the variables in the global scope (in this case ext_rew_coeef and int_rew_coeff), they will change the output of lambda. If that is not the intended behavior, it would be safer to use partial instead. 

For an example see the second answer in this thread: https://stackoverflow.com/questions/3252228/python-why-is-functools-partial-necessary